### ISSUE TITLE
preload: Error starting daemon: Devices cgroup isn't mounted

### ISSUE THREAD

**Issue by pdcastro** [balena-io/balena-cli#2250]
Preload fails with `Error starting daemon: Devices cgroup isn't mounted`
```
$ DEBUG=1 balena preload raspberrypi4-64-2.53.12+rev1-v11.9.9.img --app myApp --commit latest
...
Step 7/7 : CMD ["python3", "/usr/src/app/preload.py"]
---> Using cache
---> e5334672e585
Successfully built e5334672e585
Building Docker preloader image. [========================] 100%
...
Reading image information
Waiting for Docker to start...
Exception in thread background thread for pid 114:
Traceback (most recent call last):
File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
self.run()
File "/usr/lib/python3.6/threading.py", line 864, in run
self._target(*self._args, **self._kwargs)
File "/usr/lib/python3.6/site-packages/sh.py", line 1540, in wrap
fn(*args, **kwargs)
File "/usr/lib/python3.6/site-packages/sh.py", line 2459, in background_thread
handle_exit_code(exit_code)
File "/usr/lib/python3.6/site-packages/sh.py", line 2157, in fn
return self.command.handle_command_exit_code(exit_code)
File "/usr/lib/python3.6/site-packages/sh.py", line 815, in handle_command_exit_code
raise exc
sh.ErrorReturnCode_1:
RAN: /usr/local/bin/dockerd --storage-driver=overlay2 --data-root=/tmp/tmphaw4tmxc/docker --host=tcp://0.0.0.0:36379
STDOUT:
STDERR:
time="2021-04-13T14:36:30.248129210Z" level=warning msg="[!] DON'T BIND ON ANY IP ADDRESS WITHOUT setting --tlsverify IF YOU DON'T KNOW WHAT YOU'RE DOING [!]"
time="2021-04-13T14:36:30.248767749Z" level=info msg="libcontainerd: started new docker-containerd process" pid=130
time="2021-04-13T14:36:30Z" level=info msg="starting containerd" module=containerd revision=89623f28b87a6004d4b785663257362d1658a729 version=v1.0.0
time="2021-04-13T14:36:30Z" level=info msg="setting subreaper..." module=containerd
time="2021-04-13T14:36:30Z" level=info msg="changing OOM score to -500" module=containerd
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.content.v1.content"..." module=containerd type=io.containerd.content.v1
tim... (4360 more, please see e.stderr)
An error has occurred executing 'dockerd':
time="2021-04-13T14:36:30.248129210Z" level=warning msg="[!] DON'T BIND ON ANY IP ADDRESS WITHOUT setting --tlsverify IF YOU DON'T KNOW WHAT YOU'RE DOING [!]"
time="2021-04-13T14:36:30.248767749Z" level=info msg="libcontainerd: started new docker-containerd process" pid=130
time="2021-04-13T14:36:30Z" level=info msg="starting containerd" module=containerd revision=89623f28b87a6004d4b785663257362d1658a729 version=v1.0.0
time="2021-04-13T14:36:30Z" level=info msg="setting subreaper..." module=containerd
time="2021-04-13T14:36:30Z" level=info msg="changing OOM score to -500" module=containerd
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.content.v1.content"..." module=containerd type=io.containerd.content.v1
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.snapshotter.v1.btrfs"..." module=containerd type=io.containerd.snapshotter.v1
time="2021-04-13T14:36:30Z" level=warning msg="failed to load plugin io.containerd.snapshotter.v1.btrfs" error="path /tmp/tmphaw4tmxc/docker/containerd/daemon/io.containerd.snapshotter.v1.btrfs must be a btrfs filesystem to be used with the btrfs snapshotter" module=containerd
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.snapshotter.v1.overlayfs"..." module=containerd type=io.containerd.snapshotter.v1
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.metadata.v1.bolt"..." module=containerd type=io.containerd.metadata.v1
time="2021-04-13T14:36:30Z" level=warning msg="could not use snapshotter btrfs in metadata plugin" error="path /tmp/tmphaw4tmxc/docker/containerd/daemon/io.containerd.snapshotter.v1.btrfs must be a btrfs filesystem to be used with the btrfs snapshotter" module="containerd/io.containerd.metadata.v1.bolt"
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.differ.v1.walking"..." module=containerd type=io.containerd.differ.v1
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.gc.v1.scheduler"..." module=containerd type=io.containerd.gc.v1
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.grpc.v1.containers"..." module=containerd type=io.containerd.grpc.v1
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.grpc.v1.content"..." module=containerd type=io.containerd.grpc.v1
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.grpc.v1.diff"..." module=containerd type=io.containerd.grpc.v1
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.grpc.v1.events"..." module=containerd type=io.containerd.grpc.v1
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.grpc.v1.healthcheck"..." module=containerd type=io.containerd.grpc.v1
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.grpc.v1.images"..." module=containerd type=io.containerd.grpc.v1
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.grpc.v1.leases"..." module=containerd type=io.containerd.grpc.v1
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.grpc.v1.namespaces"..." module=containerd type=io.containerd.grpc.v1
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.grpc.v1.snapshots"..." module=containerd type=io.containerd.grpc.v1
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.monitor.v1.cgroups"..." module=containerd type=io.containerd.monitor.v1
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.runtime.v1.linux"..." module=containerd type=io.containerd.runtime.v1
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.grpc.v1.tasks"..." module=containerd type=io.containerd.grpc.v1
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.grpc.v1.version"..." module=containerd type=io.containerd.grpc.v1
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.grpc.v1.introspection"..." module=containerd type=io.containerd.grpc.v1
time="2021-04-13T14:36:30Z" level=info msg=serving... address="/var/run/docker/containerd/docker-containerd-debug.sock" module="containerd/debug"
time="2021-04-13T14:36:30Z" level=info msg=serving... address="/var/run/docker/containerd/docker-containerd.sock" module="containerd/grpc"
time="2021-04-13T14:36:30Z" level=info msg="containerd successfully booted in 0.002706s" module=containerd
time="2021-04-13T14:36:30.294869330Z" level=info msg="Graph migration to content-addressability took 0.00 seconds"
time="2021-04-13T14:36:30.294998338Z" level=warning msg="Your kernel does not support cgroup memory limit"
time="2021-04-13T14:36:30.295005303Z" level=warning msg="Unable to find cpu cgroup in mounts"
time="2021-04-13T14:36:30.295010050Z" level=warning msg="Unable to find blkio cgroup in mounts"
time="2021-04-13T14:36:30.295014370Z" level=warning msg="Unable to find cpuset cgroup in mounts"
time="2021-04-13T14:36:30.295034023Z" level=warning msg="mountpoint for pids not found"

Error starting daemon: Devices cgroup isn't mounted

Traceback (most recent call last):
File "/usr/src/app/preload.py", line 954, in <module>
result = method(**data.get("parameters", {}))
File "/usr/src/app/preload.py", line 910, in get_image_info
get_images_and_supervisor_version()
File "/usr/src/app/preload.py", line 776, in get_images_and_supervisor_version
return _get_images_and_supervisor_version()
File "/usr/src/app/preload.py", line 736, in _get_images_and_supervisor_version
with docker_context_manager(driver, mountpoint):
File "/usr/lib/python3.6/contextlib.py", line 81, in enter
return next(self.gen)
File "/usr/src/app/preload.py", line 581, in docker_context_manager
running_dockerd = start_docker_daemon(storage_driver, docker_dir)
File "/usr/src/app/preload.py", line 550, in start_docker_daemon
raise e
File "/usr/src/app/preload.py", line 543, in start_docker_daemon
running_dockerd.wait()
File "/usr/lib/python3.6/site-packages/sh.py", line 792, in wait
self.handle_command_exit_code(exit_code)
File "/usr/lib/python3.6/site-packages/sh.py", line 815, in handle_command_exit_code
raise exc
sh.ErrorReturnCode_1:
RAN: /usr/local/bin/dockerd --storage-driver=overlay2 --data-root=/tmp/tmphaw4tmxc/docker --host=tcp://0.0.0.0:36379
STDOUT:

STDERR:
time="2021-04-13T14:36:30.248129210Z" level=warning msg="[!] DON'T BIND ON ANY IP ADDRESS WITHOUT setting --tlsverify IF YOU DON'T KNOW WHAT YOU'RE DOING [!]"
time="2021-04-13T14:36:30.248767749Z" level=info msg="libcontainerd: started new docker-containerd process" pid=130
time="2021-04-13T14:36:30Z" level=info msg="starting containerd" module=containerd revision=89623f28b87a6004d4b785663257362d1658a729 version=v1.0.0
time="2021-04-13T14:36:30Z" level=info msg="setting subreaper..." module=containerd
time="2021-04-13T14:36:30Z" level=info msg="changing OOM score to -500" module=containerd
time="2021-04-13T14:36:30Z" level=info msg="loading plugin "io.containerd.content.v1.content"..." module=containerd type=io.containerd.content.v1
tim... (4360 more, please see e.stderr)
```

System:
```
$ cat /etc/os-release; docker version
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://www.archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://bugs.archlinux.org/"
LOGO=archlinux

Client:
Version: 20.10.6
API version: 1.41
Go version: go1.16.3
Git commit: 370c28948e
Built: Mon Apr 12 14:10:41 2021
OS/Arch: linux/amd64
Context: default
```

**Comment by pdcastro** [comment]
There's a suspicion that this issue might be related to the cgroups V2 feature of Docker v20.10:
https://www.docker.com/blog/introducing-docker-engine-20-10/
in which case, downgrading Docker might be a workaround.

**Comment by pdcastro** [comment]
A question was whether this issue was related to the Docker installation on Arch Linux, or related to the Docker version: v20.10.6. The following experiment using [Multipass](https://multipass.run/) to launch an Ubuntu v20.04 LTS virtual machine and then installing Docker v20.10.6 in the VM could **not** reproduce the error, suggesting that the issue is not the version of Docker:

```
multipass launch -n testvm -m 4G -d 50G -c 4 20.04
multipass exec testvm bash
curl -fsSL get.docker.com | sudo sh
sudo usermod -aG docker ubuntu
exit # so that `usermod` takes effect
```
```
multipass transfer ~/Downloads/balena-cloud-test-rpi-raspberrypi4-64-2.73.1+rev1-dev-v12.4.6.img.zip testvm:
multipass exec testvm bash
$ docker version
Client: Docker Engine - Community
 Version:           20.10.6
 ...
Server: Docker Engine - Community
  Version:          20.10.6

curl -LO https://github.com/balena-io/balena-cli/releases/download/v12.44.10/balena-cli-v12.44.10-linux-x64-standalone.zip
sudo apt-get install unzip
unzip balena-cli-v12.44.10-linux-x64-standalone.zip
unzip balena-cloud-test-rpi-raspberrypi4-64-2.73.1+rev1-dev-v12.4.6.img.zip

# sample trivial app to preload: balena idling
git clone https://github.com/balena-io-examples/balena-idling.git
./balena-cli/balena login # suggestion: token auth
./balena-cli/balena deploy test-rpi --build -s balena-idling
./balena-cli/balena preload balena-cloud-test-rpi-raspberrypi4-64-2.73.1+rev1-dev-v12.4.6.img -a test-rpi -c current --debug
```

Preload executed successfully in the last step.

**Comment by pdcastro** [comment]
Another data point: my experiment above also works with Ubuntu v20.10 (in addition to v20.04 LTS and v18.04 LTS).

**Comment by pdcastro** [comment]
This issue should be resolved in CLI v12.44.12 or later.
